libvast/src/sketch/bloom_filter.cpp

```diff
@@ -0,0 +1,96 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/sketch/bloom_filter.hpp"
+
+#include "vast/error.hpp"
+
+#include <fmt/format.h>
+
+namespace vast::sketch {
+
+frozen_bloom_filter::frozen_bloom_filter(chunk_ptr table) noexcept
+  : table_{std::move(table)} {
+  VAST_ASSERT(table_ != nullptr);
+  auto root = fbs::GetBloomFilter(table_->data());
+  auto err = unpack(*root, view_);
+  VAST_ASSERT(!err);
+}
+
+bool frozen_bloom_filter::lookup(uint64_t digest) const noexcept {
+  return view_.lookup(digest);
+}
+
+bloom_filter_params frozen_bloom_filter::parameters() const noexcept {
+  return view_.params;
+}
+
+size_t mem_usage(const frozen_bloom_filter& x) noexcept {
+  return mem_usage(x.view_), x.table_->size();
+}
+
+caf::expected<bloom_filter> bloom_filter::make(bloom_filter_config cfg) {
+  if (auto params = evaluate(cfg)) {
+    if (params->k == 0)
+      return caf::make_error(ec::invalid_argument, "need >= 1 hash functions");
+    if (params->m == 0)
+      return caf::make_error(ec::invalid_argument, "size cannot be 0");
+    return bloom_filter{*params};
+  }
+  return caf::make_error(ec::invalid_argument, "failed to evaluate parameters");
+}
+
+void bloom_filter::add(uint64_t digest) noexcept {
+  view_.add(digest);
+}
+
+bool bloom_filter::lookup(uint64_t digest) const noexcept {
+  return view_.lookup(digest);
+}
+
+const bloom_filter_params& bloom_filter::parameters() const noexcept {
+  return view_.params;
+}
+
+size_t mem_usage(const bloom_filter& x) {
+  return mem_usage(x.view_) + sizeof(x.bits_) + x.bits_.size() * 8;
+}
+
+caf::expected<frozen_bloom_filter> freeze(const bloom_filter& x) {
+  VAST_ASSERT(x.parameters().m > 0);
+  VAST_ASSERT(x.parameters().m & 1);
+  constexpr auto fixed_size = 56;
+  const auto bitvector_size = (x.parameters().m + 63) / 64;
+  const auto expected_size = fixed_size + bitvector_size * sizeof(uint64_t);
+  // FlatBuffers <= 1.11 does not correctly use '::flatbuffers::soffset_t' over
+  // 'soffset_t' in FLATBUFFERS_MAX_BUFFER_SIZE.
+  using ::flatbuffers::soffset_t;
+  if (expected_size >= FLATBUFFERS_MAX_BUFFER_SIZE)
+    return caf::make_error(
+      ec::invalid_argument,
+      fmt::format("frozen size {} exceeds max flatbuffer size of {} bytes",
+                  expected_size, FLATBUFFERS_MAX_BUFFER_SIZE));
+  // We know the exact size, so we reserve it to avoid re-allocations.
+  flatbuffers::FlatBufferBuilder builder{expected_size};
+  auto bloom_filter_offset = pack(builder, x.view_);
+  if (!bloom_filter_offset)
+    return bloom_filter_offset.error();
+  builder.Finish(*bloom_filter_offset);
+  auto buffer = builder.Release();
+  VAST_ASSERT(buffer.size() == expected_size);
```

Comment:
@dominiklohmann suggested to find out the exact size once by looking at the generated flatbuffer meta data, and then use the value here. Apparently this is also done in the `type` flatbuffer.

That said, I agree it's brittle. We'd have to update it every time we make changes to the flatbuffer. I have no strong feelings in removing it at the cost of an extra allocation. @dominiklohmann, want to chime in?

Reply:
I don't think it's brittle at all. The assertion will only trigger if you change the FlatBuffers table in a way that adds additional fields. Other than that the 56 bytes are the overhead occupied by offset tables and the remaining fields, which have a fixed size.

How often do you want to change this table? Likely almost never, and then you could adjust the 56 to what's necessary, and you will notice it immediately in CI because the assertion will trigger.

If you want to allocate a vector of almost 2 GiB size you'd surely want to call reserve on it, too, and that is exactly what we're doing here. You don't want to have to re-allocate and move around this much memory very often.

Reply:
That's why I'm asking if its part of the flatbuffers ABI. Could it happen that the number of bytes changes with e.g. an upgrade to flatbuffers 2.0.1?

Reply:
Afaik FlatBuffers guarantees that their data format is stable. FlatBuffers 2.0 is only a breaking change because there were some breaking API changes in some languages.

Reply:
@dominiklohmann can you write documentation somewhere that we can point to for these magic numbers? It seems there are multiple places in the code now that do it this way. It would make sense to treat this uniformly.

I'm not against demoting the assertion to an error, but if we decide on this, we should do it everywhere.

Reply:
The "magic number" is simply the size of the generated FlatBuffers table. It's fixed because the on-disk layout of the FlatBuffers table (assuming it is a root table) is fixed.

Is the `3` in this example a magic number?

```
auto v = vector<int>{};
v.reserve(3); // magic number?
v.push_back(a);
v.push_back(b);
v.push_back(c);
```

> It would make sense to treat this uniformly.

The numbers depend on the FlatBuffers table's data layout. You can calculate them on a 1-by-1 basis, but it'll be nearly impossible to do this generically. It's also only really worth it for larger tables where saving reallocations is worth it, e.g., for record type, segment, or bloom filters.

Reply:
Do you have a link where flatbuffer state that they guarantee that the output will remain unchanged between versions if you compile with the same input? Note that this is stronger than just binary format stability, since e.g. `flatcc` conforms to the same format but still produces different bytes as output.

However, adding these

> The "magic number" is simply the size of the generated FlatBuffers table. It's fixed because the on-disk layout of the FlatBuffers table (assuming it is a root table) is fixed.

> The assertion should only trigger if you change the FlatBuffers table in a way that adds additional fields.

> Likely almost never, and then you could adjust the 56 to what's necessary, and you will notice it immediately in CI because the assertion will trigger.

> If you want to allocate a vector of almost 2 GiB size you'd surely want to call reserve on it, too, and that is exactly what we're doing here. You don't want to have to re-allocate and move around this much memory very often.

as context would probably add enough context so that if it blows up in the future, the person seeing it can efficiently diagnose the error and update the assertion.

Reply:
I'd assume that this size is probably not guaranteed to remain stable across multiple versions or implementations of flatbuffers. But even then, it is probably worth it to do this and implement a compile time table if the assertion ever hits us.

Reply:
It has to be stable; otherwise there's a breaking change in the FlatBuffers specification.